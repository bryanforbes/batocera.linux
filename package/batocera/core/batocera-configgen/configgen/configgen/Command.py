from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
