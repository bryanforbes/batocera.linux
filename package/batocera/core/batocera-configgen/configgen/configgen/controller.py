from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
