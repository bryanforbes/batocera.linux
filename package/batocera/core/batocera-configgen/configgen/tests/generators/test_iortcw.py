from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

import pytest

from configgen.batoceraPaths import ROMS
from configgen.config import SystemConfig
from configgen.generators.iortcw.iortcwGenerator import IORTCWGenerator
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controllers
    from configgen.Emulator import Emulator


class TestIORTCWGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[IORTCWGenerator]:
        return IORTCWGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'iortcw'

    @pytest.fixture
    def emulator(self) -> str:
        return 'iortcw'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.create_dir(ROMS / 'iortcw' / 'main')

        return fs

    def test_get_in_game_ratio(self, generator: IORTCWGenerator) -> None:  # pyright: ignore
        assert generator.getInGameRatio(SystemConfig({}), {'width': 0, 'height': 0}, Path()) == 16 / 9

    def test_generate(
        self,
        generator: IORTCWGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                Path(),
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (ROMS / 'iortcw' / 'main' / 'wolfconfig.cfg').read_text() == snapshot(name='config')

    def test_generate_existing(
        self, generator: IORTCWGenerator, fs: FakeFilesystem, mock_system: Emulator, snapshot: SnapshotAssertion
    ) -> None:
        fs.create_file(
            ROMS / 'iortcw' / 'main' / 'wolfconfig.cfg',
            contents="""seta r_mode "1"
seta r_noborder "2"
seta r_fullscreen "3"
seta r_allowResize "1"
seta r_centerWindow "0"
seta r_inGameVideo "0"
seta r_customheight ""480""
seta r_customwidth ""640""
seta in_joystick "0"
seta in_joystickUseAnalog "0"
bind PAD0_A ""+flip""
bind PAD0_X ""+flip""
bind PAD0_Y ""+flip""
bind PAD0_B ""+flip""
bind PAD0_LEFTSHOULDER "flip"
bind PAD0_RIGHTSHOULDER "flip"
bind PAD0_LEFTSTICK_LEFT "+flip"
bind PAD0_LEFTSTICK_RIGHT "+flip"
bind PAD0_LEFTSTICK_UP "+flip"
bind PAD0_LEFTSTICK_DOWN "+flip"
bind PAD0_RIGHTSTICK_LEFT "+flip"
bind PAD0_RIGHTSTICK_RIGHT "+flip"
bind PAD0_RIGHTSTICK_UP "+flip"
bind PAD0_RIGHTSTICK_DOWN "+flip"
bind PAD0_LEFTTRIGGER "+flip"
bind PAD0_RIGHTTRIGGER "+flip"
seta cl_renderer "rend2"
seta r_swapInterval "1"
seta com_maxfps "10"
seta r_ext_texture_filter_anisotropic "1"
seta r_ext_max_anisotropy "8"
seta r_ext_multisample "1"
seta r_ext_framebuffer_multisample "1"
seta com_introplayed "1"
seta cl_language "3"
""",
        )
        generator.generate(
            mock_system,
            Path(),
            [],
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (ROMS / 'iortcw' / 'main' / 'wolfconfig.cfg').read_text() == snapshot

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'iortcw_api': 'rend2'},
            {'iortcw_vsync': '0'},
            {'iortcw_vsync': '1'},
            {'iortcw_fps': '120'},
            {'iortcw_filtering': '4'},
            {'iortcw_aa': '2'},
            {'iortcw_skip_video': '0'},
            {'iortcw_skip_video': '1'},
            {'iortcw_language': '1'},
        ],
        ids=str,
    )
    def test_generate_config(
        self, generator: IORTCWGenerator, mock_system: Emulator, snapshot: SnapshotAssertion
    ) -> None:
        generator.generate(
            mock_system,
            Path(),
            [],
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (ROMS / 'iortcw' / 'main' / 'wolfconfig.cfg').read_text() == snapshot
