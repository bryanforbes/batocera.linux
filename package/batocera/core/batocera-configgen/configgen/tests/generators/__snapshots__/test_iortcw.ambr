# serializer version: 1
# name: TestIORTCWGenerator.test_generate
  tuple(
    list([
      '/usr/bin/iortcw/iowolfsp',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'XDG_DATA_HOME': '/userdata/roms',
    }),
  )
# ---
# name: TestIORTCWGenerator.test_generate[config]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_aa': '2'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "2"
  seta r_ext_framebuffer_multisample "2"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_api': 'rend2'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "rend2"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_filtering': '4'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "1"
  seta r_ext_max_anisotropy "4"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_fps': '120'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "120"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_language': '1'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "1"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_skip_video': '0'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_skip_video': '1'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "1"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_vsync': '0'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_config[{'iortcw_vsync': '1'}]
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystickUseAnalog "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "1"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  
  '''
# ---
# name: TestIORTCWGenerator.test_generate_existing
  '''
  seta r_mode "-1"
  seta r_noborder "1"
  seta r_fullscreen "1"
  seta r_allowResize "0"
  seta r_centerWindow "1"
  seta r_inGameVideo "1"
  seta r_customheight ""1080""
  seta r_customwidth ""1920""
  seta in_joystick "1"
  seta in_joystick "1"
  bind PAD0_A ""+moveup""
  bind PAD0_X ""+movedown""
  bind PAD0_Y ""+useitem""
  bind PAD0_B ""+activate""
  bind PAD0_LEFTSHOULDER "weapnext"
  bind PAD0_RIGHTSHOULDER "weapprev"
  bind PAD0_LEFTSTICK_LEFT "+moveleft"
  bind PAD0_LEFTSTICK_RIGHT "+moveright"
  bind PAD0_LEFTSTICK_UP "+forward"
  bind PAD0_LEFTSTICK_DOWN "+back"
  bind PAD0_RIGHTSTICK_LEFT "+left"
  bind PAD0_RIGHTSTICK_RIGHT "+right"
  bind PAD0_RIGHTSTICK_UP "+lookup"
  bind PAD0_RIGHTSTICK_DOWN "+lookdown"
  bind PAD0_LEFTTRIGGER "+speed"
  bind PAD0_RIGHTTRIGGER "+attack"
  seta cl_renderer "opengl1"
  seta r_swapInterval "0"
  seta com_maxfps "60"
  seta r_ext_texture_filter_anisotropic "0"
  seta r_ext_max_anisotropy "2"
  seta r_ext_multisample "0"
  seta r_ext_framebuffer_multisample "0"
  seta com_introplayed "0"
  seta cl_language "0"
  seta in_joystickUseAnalog "1"
  
  '''
# ---
# name: TestIORTCWGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'menu': 'KEY_ESC',
      'pause': 'KEY_ESC',
      'restore_state': 'KEY_F9',
      'save_state': 'KEY_F5',
    }),
    'name': 'iortcw',
  })
# ---
